projects/cds/src/merchandising/cms-components/directives/attributes/attributes.directive.ts

```diff
@@ -0,0 +1,40 @@
+import {
+  Directive,
+  ElementRef,
+  Input,
+  OnChanges,
+  Renderer2,
+} from '@angular/core';
+
+@Directive({
+  selector: '[cxAttributes]',
+})
+export class AttributesDirective implements OnChanges {
+  @Input() cxAttributes: Map<string, string>;
+
+  private _attributesNamePrefix: string;
+  @Input() set cxAttributesNamePrefix(attributesNamePrefix: string) {
+    this._attributesNamePrefix = attributesNamePrefix;
+  }
+
+  constructor(private renderer: Renderer2, private elementRef: ElementRef) {}
```

Comment:
Is there a way to add attributes to an HTML element without using `Renderer2`? This is quite a low-level API that we'd like to avoid using, if possible.

Reply:
From searching around I could not find many examples of adding attributes with dynamic names other than using `Render2` or another very low level API (I cannot remember the name, but it was mentioned that it bypasses XSS protection so was not recommended to be used).

Any input here would be greatly appreciated 🙂. As you may be able to tell from our PRs, we're mainly back-end devs so are relatively naïve when it comes to many aspects of front-end dev

Reply:
Ok, let's leave it like this in that case, and maybe think about the alternative for the refactoring.
Just to double check, `Renderer2` performs encoding to prevent XSS?

Reply:
I am not sure about the low level implementation of `Renderer2`, but the [docs for `ElementRef`](https://angular.io/api/core/ElementRef#security-risk) (check the red section at the bottom) suggest using `Renderer2` as it is "safe". It is also mentioned in your contribution guidelines, but does not explicitly state that is protects against XSS (see the [DOM Manipulation section of the guidelines](https://sap.github.io/cloud-commerce-spartacus-storefront-docs/coding-guidelines/))